x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/PlanStreamInput.java

```diff
@@ -23,4 +26,10 @@ public interface PlanStreamInput {
      */
     String sourceText();
 
+    /**
+     * Read a {@link NameId} from the stream. Name-id have to be consistently
+     * mapped during serialization but need to be globally unique. This handles
+     * that combination of constraints.
+     */
+    NameId readNameId() throws IOException;
```

Comment:
Hm, we could instead have a `NameId mapNameId(long)`; that would allow us to have a method directly on `NameId`, similarly to `Source`'s :

```java
public class NameId implements Writable {
...

public static <S extends StreamInput & PlanStreamInput> Source readFrom(S in) throws IOException {
        long unmappedId = in.readLong();
        return in.mapNameId(unmappedId);
    }

}
```

Then we could get rid of the potentially confusing comment on `NameId` ("use `PlanStreamInput#readNameId` to deserialize")

I think the implementation for `mapNameId` could just use the `NameIdMapper` that we end up creating whenever we instantiate a `PlanStreamInput`. Or, even simpler, we could throw away the `NameIdMapper` class entirely and just have `PlanStreamInput` use a simple `HashMap` (which `NameIdMapper` just wraps, anyway).

Reply:
I didn't do that, but I probably should. I'll try that.